File: py/src/piper_py/detect.py
from __future__ import annotations

import os
from pathlib import Path

EXCLUDE_DIRS = frozenset(
    (
        ".venv",
        ".devenv",
        ".direnv",
        "node_modules",
        "dist",
        "build",
        ".next",
        "__pycache__",
        ".git",
        "tests",
        "test",
        "__tests__",
    )
)


def _prune_excluded(dirnames: list[str]) -> None:
    for dirname in list(dirnames):
        if dirname in EXCLUDE_DIRS:
            dirnames.remove(dirname)


def has_files(extensions: set[str], root: str = ".") -> bool:
    for dirpath, dirnames, filenames in os.walk(root):
        _prune_excluded(dirnames)
        for filename in filenames:
            if any(filename.endswith(ext) for ext in extensions):
                return True
    return False


def has_config_section(filename: str, section: str, root: str = ".") -> bool:
    filepath = Path(root) / filename
    if not filepath.exists():
        return False
    return section in filepath.read_text()

File: ts/src/detect.ts
import * as fs from "node:fs";
import * as path from "node:path";

const EXCLUDE_DIRS = new Set([
  ".venv",
  ".devenv",
  ".direnv",
  "node_modules",
  "dist",
  "build",
  ".next",
  "__pycache__",
  ".git",
  "tests",
  "test",
  "__tests__",
]);

export function hasFiles(extensions: string[], root: string = "."): boolean {
  function walk(dir: string): boolean {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return false;
    }
    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!EXCLUDE_DIRS.has(entry.name)) {
          if (walk(path.join(dir, entry.name))) return true;
        }
      } else if (extensions.some((ext) => entry.name.endsWith(ext))) {
        return true;
      }
    }
    return false;
  }
  return walk(root);
}

export function hasConfigFile(filename: string, root: string = "."): boolean {
  return fs.existsSync(path.join(root, filename));
}
